pyomo/util/latex_printer.py

```diff
@@ -0,0 +1,1336 @@
+        constraints = []
+        expressions = []
+        templatize_fcn = templatize_constraint
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, pyo.Constraint):
+        objectives = []
+        constraints = [pyomo_component]
+        expressions = []
+        templatize_fcn = templatize_constraint
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, pyo.Expression):
+        objectives = []
+        constraints = []
+        expressions = [pyomo_component]
+        templatize_fcn = templatize_expression
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, (ExpressionBase, pyo.Var)):
+        objectives = []
+        constraints = []
+        expressions = [pyomo_component]
+        templatize_fcn = templatize_passthrough
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, _BlockData):
+        objectives = [
+            obj
+            for obj in pyomo_component.component_data_objects(
+                pyo.Objective,
+                descend_into=True,
+                active=True,
+                sort=SortComponents.deterministic,
+            )
+        ]
+        constraints = [
+            con
+            for con in pyomo_component.component_objects(
+                pyo.Constraint,
+                descend_into=True,
+                active=True,
+                sort=SortComponents.deterministic,
+            )
+        ]
+        expressions = []
+        templatize_fcn = templatize_constraint
+
+    else:
+        raise ValueError(
+            "Invalid type %s passed into the latex printer"
+            % (str(type(pyomo_component)))
+        )
+
+    if isSingle:
+        temp_comp, temp_indexes = templatize_fcn(pyomo_component)
+        variableList = []
+        for v in identify_components(
+            temp_comp, [ScalarVar, _GeneralVarData, IndexedVar]
+        ):
+            if isinstance(v, _GeneralVarData):
+                v_write = v.parent_component()
+                if v_write not in ComponentSet(variableList):
+                    variableList.append(v_write)
+            else:
+                if v not in ComponentSet(variableList):
+                    variableList.append(v)
+
+        parameterList = []
+        for p in identify_components(
+            temp_comp, [ScalarParam, _ParamData, IndexedParam]
+        ):
+            if isinstance(p, _ParamData):
+                p_write = p.parent_component()
+                if p_write not in ComponentSet(parameterList):
+                    parameterList.append(p_write)
+            else:
+                if p not in ComponentSet(parameterList):
+                    parameterList.append(p)
+
+        # Will grab the sets as the expression is walked
+        setList = []
+
+    else:
+        variableList = [
```

Comment:
We should reconsider this mode.  The writers have been moving *away* from pre-gathering variables / parameters / sets and instead just recording the variables as they are encountered when walking expressions.  In part, this is because we are relaxing *where* a Var has to exist (in particular, it doesn't have to exist within the block where is it used).